Sessions: Save & display date using the correct timezone

ryelle opened the pull request:
This PR updates session date (`_wcpt_session_time`) displays to output a timezone-aware date. It also replaces `strtotime` with `date_create` when saving the timestamp, to generate the correct timestamp value. Lastly, it removes the (now unnecessary) timezone computation from the Live Schedule block.

Fixes #226 

This will need to be merged right before running the session time fixer script (see https://github.com/WordPress/wordcamp.org/issues/226#issuecomment-577916924), but this means sites will display the wrong time until all are fixed. On the other hand, waiting until after will have the sites that get fixed displaying incorrectly as the script runs.  Thoughts❓

Edit: I think the script will take about 10-15min to run, so it's not a large window either way.

### How to test the changes in this Pull Request:

1. Set a timezone on your site, if not done already
2. Add/edit some sessions with times
3. Check that the right value is saved:
    - `wp post meta get <id> _wcpt_session_time`
    - Go to `https://www.unixtimeconverter.io/[value]`
    - It should show you the timestamp value in UTC & your timezone, unless you set the site timezone to UTC+0, the UTC version _should not_ be the session time
4. Check that the date displays correctly…
    - in the datepicker for the session
    - in the session list table
    - in the schedule shortcode
    - anywhere else a human-readable date is output
5. Check that the live schedule is calculating correctly (see #324 for LS test instructions)

ryelle committed:
Sessions: Fix date display to use site timezone

This updates everywhere the session time is displayed, to display the date using the site's timezone. Previously, the date was displayed using UTC everywhere, which didn't matter because it was consistent. Now that we're using dates in javascript, this is getting messy (@wordpress/date expects to use the site timezone).

See https://make.wordpress.org/core/2019/09/23/date-time-improvements-wp-5-3/

ryelle committed:
Sessions: Save date & time using the real (local) timestamp

The session date & time is saved as a unix timestamp, but previously was saved using the UTC timezone, meaning a session at 1pm on a site in New York would actually be saved as being at 8am EST. Now that dates are displayed using the site timezone, this issue is obvious — so we need to create the timestamp in a timezone-aware way.

ryelle committed:
Remove unnecessary timezone calculations

Now that the session time is the correct time, we don't need to add in a reverse offset calculation.

ryelle committed:
Fix phpcs for `date`

coreymckrill reviewed (approved):
👍 All the tests worked on my local environment... after I remembered to rebuild the `wordcamp-blocks` scripts 🤦‍♂ 

> I think the script will take about 10-15min to run, so it's not a large window either way.

Yeah, as long as we do it on a day when there aren't any WordCamps happening, it shouldn't be a big deal.

iandunn commented:
Great effort here so far, kudos 👍 

How will these external clients be affected?

* the Events Widget
* make.w.org/community/events (probably the same as ^, but worth checking)
* https://github.com/wordpress-mobile/WordCamp-Android
* https://github.com/ctarda/WordCamp (probably the same as ^, but worth checking)

At first glance it seems like this would break back-compat with all those API clients, is that right? If so, a safer approach might be to:

* keep the existing code and values, but treat `_wcpt_session_time` as deprecated
* save the new values as something like `_wcpt_session_time_with_tz`
* update all wordcamp.org code to use the new field
* update api.w.org/events to also serve `_wcpt_session_time_with_tz`, but continue serving `_wcpt_session_time`
* if there are any Events Widget bugs related to TZ, then update it to use `_wcpt_session_time_with_tz`. older versions would still always need `_wcpt_session_time` to remain in the API though.
* add lots of documentation in code, api responses, written docs, etc that `_wcpt_session_time` is deprecated and buggy

ryelle commented:
> How will these external clients be affected?

Most of those clients don't use **session times.** The events widget & make.w.org page only seem to be showing times for meetup.com events, not individual WordCamp sessions. The iOS WordCamp app doesn't show session info. It _would_ affect the android app— but it seems [they're still on the v1 endpoints,](https://github.com/wordpress-mobile/WordCamp-Android/blob/9a7f60b96017114a02d6dcca264855513713229e/app/src/main/java/org/wordcamp/android/networking/WPAPIClient.java#L16-L24) so I don't know if we need to guarantee compatibility. On the other hand, I'm fine with dropping in a filter so that v1 requests for `_wcpt_session_time` are actually served `_wcpt_session_time_utc`.

iandunn commented:
You're right, those were bad examples. It seems like back-compat is still an issue, though?


_( Note for our future selves: If we do [make similar changes for the WCPT start date/time](https://github.com/WordPress/wordcamp.org/issues/226#issuecomment-580368588) then it seems like that would affect the Events Widget, etc, since in some cases the date would be shifted by a day. )_



> _wcpt_session_time_utc

That's a much better name 👍

ryelle commented:
I dug a little more and was unable to find an active project that uses the sessions API (using github's global code search). I'll acknowledge that there might still be projects that use this value, but I think the added technical debt of defaulting to the incorrect value outweighs any potential impact of changing this. So, here's what I'm planning on doing:

- Update the v1 endpoint to continue using the incorrect value, calculated from the real value. This way, the WordCamp Android app will continue to work.
- Add a parameter to the v2 endpoint to switch into “UTC mode”, which will return the incorrect value, again calculated from the real value — so anyone using the v2 endpoint can toggle this switch on while they fix their code.
- Draft something summarizing all this to post on make/meta
- Post that, then wait a few days for anyone to raise issues (can bring it up in the bug scrub on Thurs, too)
- Unless someone brings up a new issue, merge this PR & run the script (probably not until the 10th)

coreymckrill commented:
This sounds like a reasonable approach to me. This way we're not "silently" changing the API and possibly breaking backcompat for someone.

ryelle committed:
JSON API (v1): Calculate the faux-UTC value for backwards compatibility with API consumers

This API is still in use (by the WordCamp Android app), so we'll continue offering the miscalculated time value for consistency.

ryelle committed:
REST API (v2): Add a parameter to toggle on the faux-UTC time output

iandunn commented:
I definitely agree that adding more technical debt is undesirable. We have a lot of it already, and avoiding/minimizing it is important.

I'm still really concerned that this would break things that others have built, though. I'm not confident that a GitHub search would find enough projects.

Do you think there's a way that plan can be tweaked to avoid breaking back-compat, while still minimizing the technical debt?
